### ISSUE TITLE
matchers: add firstOf matcher

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy#20666]
For something like ALPN list matching, we need a way to say: iterate over the input, which is a list, until the inner matches accepts.
The core API does not provision this use case, so this has to be done as a meta-matcher with a custom matcher.

**Comment by zhxie** [comment]
I think it is a good idea to add `firstOf` matcher, but first we have to reconstruct `Envoy::Matcher::DataInput`. We can only get string result currently, and we need list for ALPN and `ProtobufWkt::Value` for metadata.

**Comment by kyessenov** [comment]
Correct. We can make `DataInput` produce a variant of a string and a vector of strings like `CelValue` does it as a start. Protobuf is not good for generic runtime objects AFAIK. Alternatively, we can use string and then parse as a list when necessary yielding string views.
